icicle/backend/cpu/include/ntt_cpu.h

```diff
@@ -65,56 +61,169 @@ namespace ntt_cpu {
+            if (hierarchy_0_layer_idx == 0) {
+              nof_blocks = 1 << ntt_data.ntt_sub_hierarchies.hierarchy_0_layers_sub_logn[hierarchy_1_layer_idx][2];
+              nof_subntts = 1 << ntt_data.ntt_sub_hierarchies.hierarchy_0_layers_sub_logn[hierarchy_1_layer_idx][1];
+            } else if (hierarchy_0_layer_idx == 1) {
+              nof_blocks = 1 << ntt_data.ntt_sub_hierarchies.hierarchy_0_layers_sub_logn[hierarchy_1_layer_idx][2];
+              nof_subntts = 1 << ntt_data.ntt_sub_hierarchies.hierarchy_0_layers_sub_logn[hierarchy_1_layer_idx][0];
+            } else {
+              nof_blocks = 1
+                           << (ntt_data.ntt_sub_hierarchies.hierarchy_0_layers_sub_logn[hierarchy_1_layer_idx][0] +
+                               ntt_data.ntt_sub_hierarchies.hierarchy_0_layers_sub_logn[hierarchy_1_layer_idx][1]);
+              nof_subntts = 1;
+            }
+            nof_blocks = nof_blocks >> 1;
+            size_t num_chunks = (std::thread::hardware_concurrency()) << 5; // Adjust based on the number of threads
+            size_t chunk_size =
+              (nof_blocks * nof_subntts * nof_hierarchy_1_subntts_todo_in_parallel + num_chunks - 1) / num_chunks;
+            for (size_t i = 0; i < num_chunks; ++i) {
+              size_t start_index = i * chunk_size;
+              size_t end_index =
+                std::min(start_index + chunk_size, nof_blocks * nof_subntts * nof_hierarchy_1_subntts_todo_in_parallel);
+              taskflow.emplace([this, start_index, end_index, nof_blocks, hierarchy_1_layer_idx,
+                                hierarchy_1_subntts_chunk_idx, hierarchy_0_layer_idx, nof_subntts,
+                                nof_hierarchy_1_subntts_todo_in_parallel]() {
+                for (uint32_t j = start_index; j < (end_index); j++) {
+                  uint32_t hierarchy_1_subntt_idx_in_chunk = j / (nof_subntts * nof_blocks);
+                  uint32_t hierarchy_0_block_idx_half = (j / nof_subntts) % nof_blocks;
+                  uint32_t hierarchy_0_block_idx = hierarchy_0_block_idx_half << 1;
+                  uint32_t hierarchy_0_subntt_idx = j % nof_subntts;
+                  NttTaskCoordinates ntt_task_coordinates(
+                    hierarchy_1_layer_idx,
+                    hierarchy_1_subntts_chunk_idx * nof_hierarchy_1_subntts_todo_in_parallel +
+                      hierarchy_1_subntt_idx_in_chunk,
+                    hierarchy_0_layer_idx, hierarchy_0_block_idx, hierarchy_0_subntt_idx, false);
+                  NttTask<S, E> task(ntt_task_coordinates, ntt_data);
+                  task.execute();
+                  ntt_task_coordinates.hierarchy_0_block_idx = hierarchy_0_block_idx + 1;
+                  NttTask<S, E> task_with_elements_in_the_same_cachline(ntt_task_coordinates, ntt_data);
+                  task_with_elements_in_the_same_cachline.execute();
+                }
+              });
+            }
+            executor.run(taskflow).wait();
```

Comment:
do you have to wait here due to dependency? If yes, let's leave a todo to see if we can avoid it

Reply:
Added TODO in this instance